workflow-ui/core-android/src/main/java/com/squareup/workflow/ui/WorkflowViewStub.kt

```diff
@@ -29,57 +30,115 @@ import android.view.ViewGroup
  *
  * ```
  *     class YourLayoutRunner(view: View) {
- *       private val child = view.findViewById<WorkflowViewStub>(R.id.child_stub)
+ *       private val childStub = view.findViewById<WorkflowViewStub>(R.id.child_stub)
+ *
+ *       // Totally optional, since this view is also accessible as [childStub.actual].
+ *       // Note that R.id.child was set in XML via the square:inflatedId parameter.
+ *       private val child: View by lazy {
+ *         view.findViewById<View>(R.id.child)
+ *       }
  *
  *       override fun showRendering(
  *          rendering: YourRendering,
  *          viewEnvironment: ViewEnvironment
  *       ) {
- *         child.update(rendering.childRendering, viewEnvironment)
+ *         childStub.update(rendering.childRendering, viewEnvironment)
  *       }
  *     }
  * ```
+ *
+ * **NB**: If you're using a stub in a `RelativeLayout` or `ConstraintLayout`, relationships
+ * should be tied to the stub's `app:inflatedId`, not its `android:id`.
  */
 class WorkflowViewStub @JvmOverloads constructor(
   context: Context,
   attributeSet: AttributeSet? = null,
   defStyle: Int = 0,
   defStyleRes: Int = 0
 ) : View(context, attributeSet, defStyle, defStyleRes) {
+  /**
+   * On-demand access to the view created by the last call to [update],
+   * or this [WorkflowViewStub] instance if none has yet been made.
+   */
+  var actual: View = this
+    private set
+
+  /**
+   * The id to be assigned to views created by [update]. If the inflated id is
+   * [View.NO_ID] (its default value), new views keep their original ids.
+   */
+  @IdRes var inflatedId: Int
+
   init {
+    val attrs = context.obtainStyledAttributes(
+        attributeSet, R.styleable.WorkflowViewStub, defStyle, defStyleRes
+    )
+    inflatedId = attrs.getResourceId(R.styleable.WorkflowViewStub_inflatedId, NO_ID)
+    attrs.recycle()
+
     setWillNotDraw(true)
   }
 
   /**
-   * On-demand access to the delegate established by the last call to [update],
-   * or this [WorkflowViewStub] instance if  none has yet been set.
+   * Function called from [update] to replace this stub, or the current [actual],
+   * with a new view. Can be updated to provide custom transition effects.
+   *
+   * Note that this method is responsible for copying the [layoutParams][getLayoutParams]
+   * from the stub to the new view. Also note that in a [WorkflowViewStub] that has never
+   * been updated, [actual] is the stub itself.
    */
-  var actual: View = this
+  var replaceOldViewInParent: (ViewGroup, View) -> Unit = { parent, newView ->
+    val index = parent.indexOfChild(actual)
+    parent.removeView(actual)
+    actual.layoutParams
+        ?.let { parent.addView(newView, index, it) }
+        ?: run { parent.addView(newView, index) }
+  }
+
+  /**
+   * Sets the visibility of this stub as usual, and also that of [actual].
+   * Any new views created by [update] will be assigned this visibility.
+   */
+  override fun setVisibility(visibility: Int) {
```

Comment:
Can we override this as a var to be more idiomatic?

Reply:
Maybe? I'll see.

Reply:
Nope. 

https://youtrack.jetbrains.com/issue/KT-6653

Reply:
Oh well